blobs: assorted bug fixes

knz opened the pull request:
See the individual commits for details.

knz committed:
blobs: adjust the check for relative paths

The local storage code enables filename paths to target
a directory outside of the temp storage dir, but _only_
if the operator enables so using symlinks from "inside"
to "outside".

To enforce this, the code relies on `filepath.Join` to strip out /
normalize occurrences of `../`.

This patch clarifies this by adding a comment and a missing test.

Release note: None

knz committed:
blobs: properly refuse to process directories in localStorage.ReadFile()

Prior to this patch, `(*localStorage).ReadFile()` was blindly opening
the target path, assuming that the OS would report an error if the
path targets a directory.

However this assumption is incorrect, as opening a directory
for regular reading is a valid operation on many systems.

This patch fixes the issue by adding a stat call on the open file.

Release note (bug fix): CockroachDB will now more reliably produce an
error message when a client mistakenly uses a directory name instead
of a file name with `nodelocal://` in bulk I/O operations.

knz committed:
blobs, fileutil: properly handle cross-filesystem renames

Prior to this patch, the `(*localStorage) WriteFile()` method was
expecting to be able to use `os.Rename` to move a newly created file
from `$TMPDIR` to the external I/O directory.

This is bound to fail hard in most deployment scenarios, where
typically `/tmp` is on a different filesystem (most often RAM-only).

This patch fixes it by introducing a more correct `fileutil.Move()`
function and using it from the blob store.

Release note (bug fix): CockroachDB will now properly handle cases
where the system's temporary directory lives on a different filesystem
from the external I/O directory. Symptoms of the bug would be an error
when trying to export data using a `nodelocal://` URL.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/42542)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
Thanks for doing this. I don't fundamentally disagree with anything you've done in this PR but it seems like a really bad idea to be using TMPDIR here at all. 

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42542#-:-LtziZa3A9CAg8wOYz2C:b8qxc0n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @g3orgia)

___
*[pkg/blobs/local_storage.go, line 69 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42542#-LtzhxVM3BBG27_qj_Ht:-LtzhxVN52Y2zvl-iBsd:bftk3sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4801e3f79f73270d9b51f7ecf25b884955c8e64/pkg/blobs/local_storage.go#L69)):*
> ```Go
> 		return err
> 	}
> 	tmpFile, err := ioutil.TempFile(os.TempDir(), filepath.Base(fullPath))
> ```

My sense is that while handling the cross device errors is a good idea, we probably shouldn't be writing the temp file into os.TempDir() but rather we should be writing it to the external IO dir. This would eliminate cross device IO except in cases where users have subdirectories of their external IO dir on different devices. 

Writing these files to TMPDIR  will potentially cause lots of problems for our users. We expect the external io dir to be big enough to hold files but I don't expect users to anticipate that they need adequate space in their TMPDIR.
___
*[pkg/blobs/local_storage.go, line 73 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/42542#-LtziagREIvf6mIEXeVu:-LtziagREIvf6mIEXeVv:b-4x7iw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f4801e3f79f73270d9b51f7ecf25b884955c8e64/pkg/blobs/local_storage.go#L73)):*
> ```Go
> 		return errors.Wrap(err, "creating tmp file")
> 	}
> 	// TODO(georgiah): cleanup intermediate state on failure
> ```

@georgiah I'd love to see this done before your leave. 


<!-- Sent from Reviewable.io -->

ajwerner commented:
I'm sad I didn't catch the TMPDIR thing the first time around. Thanks again for investigating this and whipping up a fix so quickly.

dt reviewed (commented):
👍 on commits 2 and 3, but first commit is explicitly _not_ the intended behavior.

knz commented on pkg/blobs/local_storage.go:
```diff
@@ -39,9 +40,22 @@ func newLocalStorage(externalIODir string) (*localStorage, error) {
+	// there, or invalid path components like the special names on
+	// windows. We really want to check upfront that the path is valid.
+	// filepath.Abs() will do this for us.
+	absPath, err := filepath.Abs(localBase)
```
I'll add the missing tests and comments then. Thanks for explaining.

knz commented:
Another fun find as I am re-checking this: the current code leaks files in the temp dir if the copy fails. It forgets to remove the original file when it's done...

ajwerner commented:
> Another fun find as I am re-checking this: the current code leaks files in the temp dir if the copy fails. It forgets to remove the original file when it's done...

IIRC this was known and had a TODO.

dt commented:
nit: commit 2 headline could ideally mention what it fixed, eg. `blobs: fix ReadFile() handling of directories`

knz commented:
Q: is the WriteFile() method ever used to overwrite existing files?

bdarnell commented:
> Another fun find as I am re-checking this: the current code leaks files in the temp dir if the copy fails. It forgets to remove the original file when it's done...

The `O_TMPFILE`/`linkat(2)` pattern could be used to avoid these leaks (create an unnamed file on the target filesystem, and don't give it a name until you're done writing to it). However, it's not portable, so we'd continue to need an alternative cleanup path. It's linux-only, and needs a relatively recent kernel (3.11, which notably excludes RHEL 7).

knz commented:
I'm revisiting this code right now to put some defers in place as a 80% solution. Will ping when the PR is updated.

knz committed:
blobs: avoid using $TMPDIR + remove temp files upon I/O errors

Prior to this patch, a file upload to node-local storage would have
the following defects:

- it would attempt to upload the data first to that node's $TMPDIR,
  and only then move the file to its final desired location. This is
  defective because often in production systems the directory pointed to
  by $TMPDIR is small (e.g. a RAMdisk).

- if the I/O fails, the temporary file was not deleted.

This patch fixes the two issues by uploading the file in its final
target directory and removing the newly created temporary file if an
I/O operation fails.

Release note (bug fix): CockroachDB now avoids using `$TMPDIR` (often
set `/tmp`) during bulk I/O operations. This prevents errors occurring
when the `$TMPDIR` disk capacity is small compared to the configured
external I/O diretory.

Release note (bug fix): Temporary files created during certain bulk
I/O operations are now properly deleted when an error occurs. This
prevents left-over temporary files being retained in the system and
leaking disk usage over time.

knz commented:
Added one more commit to fix the temp file leakage, PTAL.

dt commented on pkg/blobs/local_storage.go:
```diff
@@ -37,11 +37,18 @@ func newLocalStorage(externalIODir string) (*localStorage, error) {
 func (l *localStorage) prependExternalIODir(path string) (string, error) {
 	localBase := filepath.Join(l.externalIODir, path)
-	// Make sure we didn't ../ our way back out.
+	// Make sure no symlinks accessed a way out.
```
isn't this comment the opposite of what the code / comment above is doing?

knz replied:
woops. Corrected during the rebase.

knz commented:
TFYRs!

bors r+
